metagamer/environments/tictactoe.py

```diff
@@ -0,0 +1,176 @@
+"""
+This file provides some simple tools to simulate and work with a tic tac toe board.
+The functions provided assume the board is represented by
+    - np array of shape 3,3
+    - player 1 is 1, player 2 is -1 and empty is 0
+
+
+"""
+import logging
+
+import numpy as np
+import gym
+from gym import spaces, error
+
+from typing import Iterable, Any, Tuple, Dict, List, Optional, Iterator
+
+logger = logging.getLogger(__name__)
+
+
+def turn_iterator(turn_order: Iterable[Any]) -> Iterator[Any]:
+    """Infinitely iterates through the order provided"""
+    while True:
+        for turn in turn_order:
+            yield turn
+
+
+def check_win(board: np.array) -> int:
+    """
+    Given a 3*3 numpy array whose elements are 0, 1, -1
+    representing empty, player1 and player 2,
+    return the player who's won if any or 0 if no winner found
+
+    Args:
+        board (np.array): 3 x 3 board
+
+    Returns:
+        int: 0  -> no winner
+             1  -> player 1
+             -1 -> player 2
+
+    """
+    assert board.shape == (3, 3)
+    for axis in [0, 1]:
+        axis_sum = board.sum(axis=axis)
+        if 3 in axis_sum:
+            return 1
+        elif -3 in axis_sum:
+            return -1
+
+    diag = board.diagonal().sum()
+    diag_inv = np.fliplr(board).diagonal().sum()
+    for diag_sum in [diag, diag_inv]:
+        if diag_sum == 3:
+            return 1
+        elif diag_sum == -3:
+            return -1
+
+    # no winner yet
+    return 0
+
+
+def to_one_hot(board: np.array) -> np.array:
+    """
+    Convert the representation to one hot encoding
+    We assume a 3x3 board input
+
+    Args:
+        board (np.array): 3 x 3
+
+    Returns:
+        3 x 3 x 3:
+            board[:, :  0] = location of empty squares
+            board[:, :, 1] = location of player 1
+            board[:, :, 2] = location of player 2 moves
+
+    """
+    oh = np.stack((board == 0, board == 1, board == -1))
+    return oh.shape.astype(int)
+
+
+def to_human(board: np.array) -> np.array:
+    """Convert this into a """
+    human_board = np.full(board.shape, " ")
+    human_board[np.where(board == 1)] = "X"
+    human_board[np.where(board == -1)] = "O"
+    return human_board
+
+
+class TicTacToeEnv(gym.Env):
+    """
+    TicTacToe environment in the openai gym style
+
+    This doesn't support observation_space but we can adapt that later
+
+    """
+
+    metadata = {"render.modes": ["human"]}
+
+    TURN_ORDER = [1, -1]
+
+    def __init__(self):
+        self.board_shape = 3, 3
+        self.symbols = {1: "X", -1: "O"}
+        self.action_space = spaces.Tuple([spaces.Discrete(3), spaces.Discrete(3)])
+        # we don't support this right now
+        self.observation_space = None
+
+        self.board = None
+        self.turn_iterator = None
+        self.curr_turn = None
+        self.done = None
+
+        self.reset()
```

Comment:
I wonder (and haven't taken any time yet to research) how many of these attributes are prescribed by openai, and how many are ours?

Reply:
action_space and observation_space are required.

board, turn_iterator, curr_turn, done are all internal state representation
    
board_shape and symbols are like constants. I think symbols is unused and should be removed.

Reply:
Thanks

Reply:
I've updated doc to make this clearer